go/src/oauth2client/oauth2client.go

```diff
@@ -0,0 +1,285 @@
+	authorizeHandler func(string) string
+}
+
+// Run the three-legged oauth flow, including a authorize flow and a verify
+// flow. Returns the token object.
+func (c ThreeLeggedClient) GetToken(scopes string) (*Token, error) {
+	// In the authorize flow, user will paste a verification code back to console.
+	code := authorizeFlow(c.secret, scopes, c.authorizeHandler)
+
+	// The verify flow takes in the verification code from authorize flow, sends a
+	// POST request containing the code to fetch oauth token.
+	return verifyFlow(c.secret, scopes, code)
+}
+
+// Run the two-legged oauth flow, will create a JWT token and use the JWT token
+// to fetch an oauth token. Returns the token Object
+func (c TwoLeggedClient) GetToken(scopes string) (*Token, error) {
+	// Read the private key in service account secret.
+	pemBytes := []byte(c.secret["private_key"].(string))
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, fmt.Errorf("Failed to read private key.")
+	}
+
+	// Ignore error, handle the error case below.
+	pkcs8key, _ := x509.ParsePKCS8PrivateKey(block.Bytes)
+
+	// Create a pkeyInterface object containing the private key. The
+	// pkeyInterface object has a sign function to sign a hash.
+	pkey, ok := pkcs8key.(pkeyInterface)
+	if !ok {
+		return nil, fmt.Errorf("Failed to parse private key.")
+	}
+
+	// Get the JWT token
+	jwt, err := createJWT(c.secret, scopes, pkey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct the POST request to fetch the OAuth token.
+	params := url.Values{
+		"assertion":  []string{jwt},
+		"grant_type": []string{jwtBearerUrn},
+	}
+
+	// Send the POST request
+	response, err := http.PostForm(c.secret["token_uri"].(string), params)
+	if err != nil {
+		return nil, fmt.Errorf("Failed in http request to get service account token.")
+	}
+	defer response.Body.Close()
+
+	// The oauth token will be in the response body json.
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to read http body for service account token request.")
+	}
+	var token *Token
+	if err := json.Unmarshal(body, &token); err != nil {
+		return nil, fmt.Errorf("Failed to parse json.")
+	}
+	return token, nil
+}
+
+// Create a new OAuth2 Client with default Authorize Handler
+// secretBytes: JSON text that represents either an OAuth client ID or a
+// service account.
+func NewClient(secretBytes []byte) (Client, error) {
+	return NewClientWithAuthorizeHandler(secretBytes, defaultAuthorizeFlowHandler)
+}
+
+// Create a new OAuth2 Client with given Authorize Handler
+// secretBytes: JSON text that represents either an OAuth client ID or a
+// service account.
+// authorizeHandler: a function that handles three-legged OAuth authorize flow.
+// It should take in an URL, let the user authorize access on that URL, and
+// the verify code.
+func NewClientWithAuthorizeHandler(secretBytes []byte, authorizeHandler func(string) string) (Client, error) {
```

Comment:
I suggest we don't provide 2 functions. Just have a single NewClient([]byte, func(string)string) is enough. It is trivial to pass (str, nil). It is not worth saving the extra parameter.